keps/sig-network/4770-endpointslice-controller-flexibility/README.md

```diff
@@ -0,0 +1,362 @@
+
+This will have no effect on other EndpointSlice controller implementations since they will not be influenced by the presence of this label.
+
+## Design Details
+
+### Well-Known Label
+
+The kube-controller-manager will pass to the Endpoints, EndpointSlice and EndpointSlice Mirroring Controllers an informer selecting services that are not labeled with `service.kubernetes.io/endpoint-controller-name`. Thus, if the label is added to an existing service (by updating the service), the service with the label will be considered as a deleted service for the controllers, and the Endpoints and EndpointSlices will be deleted. If a Service is created with the label, the controllers will not be informed about it, so the Endpoints and EndpointSlices will not be created. If the label is removed from an existing service (by updating the service), the service with the label will be considered as a newly created service for the controllers, and the Endpoints and EndpointSlices will be created.
+
+In the Endpoints, EndpointSlice and EndpointSlice Mirroring Controllers, the behavior to create Endpoints/EndpointSlices on service creation and the behavior to delete the Endpoints/EndpointSlices on service deletion is already in place. Only the service informer passed to these controllers must be tweaked for the proposed well-known label (`service.kubernetes.io/endpoint-controller-name`) to work properly.
+
+### Test Plan
+
+[x] I/we understand the owners of the involved components may require updates to
+existing tests to make this code solid enough prior to committing the changes necessary
+to implement this enhancement.
+
+##### Prerequisite testing updates
+
+##### Unit tests
+
+TDB
+
+##### Integration tests
+
+- Usage of `service.kubernetes.io/endpoint-controller-name` on services
+    * With the feature gate enable, a service is created with the label and the service has then no Endpoints neither EndpointSlices. Then service is updated removing the label and the service has now Endpoints and EndpointSlices.
+    * With the feature gate enable, a service is created without the label, the service has Endpoints and EndpointSlices. Then service is updated with the label and the service has no longer any Endpoints nor EndpointSlices.
+    * With the feature gate disabled, a service is created with the label and the service has EndpointSlices as it would have had without the label.
+
+##### e2e tests
+
+- Usage of `service.kubernetes.io/endpoint-controller-name` on services
+    * A service is created with the label, an EndpointSlice is manually created with an endpoint (simulating an external controller). The test verifies the service is reachable. 
+
+### Graduation Criteria
+
+#### Alpha
+
+- Feature implemented behind feature gates (`EndpointControllerNameWellKnownLabel`). Feature Gates are disabled by default.
+- Documentation provided.
+- Initial unit, integration and e2e tests completed and enabled.
+
+#### Beta
+
```

Comment:
There has to be some e2e conformance of these out of tree controllers, since is an external controller adding an endpointslice, how do we validate these out of tree controllers are generating the endpointslices accordingly ? what if these controllers does not respect the selectors and generete the endpointslices anyway?

@robscott I really like to hear your thoughts on this

Reply:
That's an interesting thought. I'd assumed that the purpose of this KEP was to build controllers that would intentionally have different behavior than the default in-tree EndpointSlice controller. In some cases that might be to select a different IP for a Pod, but in other cases that might be to select an entirely different type of resource, for example a custom backend type. 

This does certainly make it more likely that unexpected variations of EndpointSlices will exist, which is certainly something to watch out for. Since the outset of the EndpointSlice API _anyone_ has been able to create EndpointSlices, and sometimes unexpected variations have led to bugs in controllers consuming those EndpointSlices, but I don't think that's necessarily new with this KEP, just potentially a more common set of edge cases that we'll need to be sure that controller authors are aware of. 

For example, `nodeName` is optional in EndpointSlices, but some controllers likely assume it exists because the EndpointSlice controllers in k/k _generally_ populate it. The same is largely true for zone and hostname (see [podToEndpoint](https://github.com/kubernetes/kubernetes/blob/7df2b05e772c4190f53038cf529ef00769818899/staging/src/k8s.io/endpointslice/utils.go#L58-L69) for a reference).

I think this risk is unavoidable with this KEP, so I'd like to see it called out + a corresponding commitment to improved documentation around this as a requirement for graduating past alpha. Unfortunately I can't think of a way to write conformance tests that would help us here though.

/cc @swetharepakula @gauravkghildiyal

Reply:
I wonder if we want to test that EndpointSlices created before a Service still work how we want them to (kube-proxy behavior for example), even after that Service is eventually added?